### DISCUSSION TITLE
Connectorx should take less time to execute sql queries but it is taking more time than sqlite3 module

### DISCUSSION THREAD

**Discussion by deepakpunia20** [sfu-db/connector-x#251] [Q&A]
Hi @wangxiaoying ,

Connectorx should take less time to execute sql queries but it is taking more time than sqlite3 module. Below is the example:

![image](https://user-images.githubusercontent.com/73750090/157375335-4730a74f-5e90-4d41-bc9d-2c3e2c5e6be0.png)

![image](https://user-images.githubusercontent.com/73750090/157375377-ac9536ee-2c52-44b4-8278-5ccb9f8bf3a1.png)

**Comment by wangxiaoying** [comment]
Hi @deepakpunia20 ,

ConnectorX is mainly targeting on the large query result fetching scenario. It speeds up the process by optimizing the client-side execution and saturating both network and machine resource through parallelism. When query execution is the bottleneck (for example the result size is small in your case, or the query is very complex), there will be overhead coming from metadata fetching. In ConnectorX, there are up to three info that will be fetched before issue the query to database:

1. MIN, MAX query for partition range (if partition is applied)
2. COUNT query (if `return_type=pandas`)
3. schema fetching query

In your situation the overhead comes from 2 and 3. In order to avoid the potentially costly COUNT query, we suggest to use Arrow as an intermediate destination from ConnectorX and convert it into Pandas using Arrow’s [to_pandas API](https://arrow.apache.org/docs/python/generated/pyarrow.Table.html?pyarrow.Table.to_pandas). For example:

```Python
import connectorx as cx

table = cx.read_sql(db_uri, query, return_type="arrow")
df = table.to_pandas(split_blocks=False, date_as_object=False)
```

Please feel free to have a try. It may reduce the time a bit. But since the query result in your case is too small, the overhead from 3 may still affect the end-to-end time.

**Reply by deepakpunia20** [reply]
Hi @wangxiaoying ,

I tried using arrow, though it reduced time in comparison to connectorx but still it's taking more time than sqlite3 module. Do you have way to reduce selection time from large dataset(12GB size) ? Generally source datasets(databases) are of large size, So our emphasis should be to select and update data in such large datasets quickly rather than focusing on result dataset because results datasets are comparatively very small in size.


![image](https://user-images.githubusercontent.com/73750090/157466384-f3f9b860-6a51-424e-8542-129dc7612ea0.png)

![image](https://user-images.githubusercontent.com/73750090/157466448-302b73c7-4d15-46ec-8846-3d912430a91e.png)

**Reply by wangxiaoying** [reply]
ConnectorX is targeting on optimizing large query result in which **converting the data from binary to pandas on the client machine** is the bottleneck. I think you can try partition the query either manually or automatically on a numerical column like examples [here](https://github.com/sfu-db/connector-x#examples), so that we will issue subqueries in parallel. But in general, when query result is very small like in this case, there is very little that we can accelerate on the client side, and the overhead we have for accelerating large result may become obvious.
